src/badguy/root.cpp

```diff
@@ -16,79 +16,290 @@
+      break;
+    default: assert(false); break;
+  }
+
+  const float gravity = Sector::get().get_gravity() * 100.f;
+  for (int i = 0; i < 5; i++)
+  {
+    const Vector velocity(graphicsRandom.randf(-100, 100),
+                          graphicsRandom.randf(-400, -300));
+    Sector::get().add<SpriteParticle>("images/particles/corrupted_rock.sprite",
+                                      "piece-" + std::to_string(i),
+                                      basepos, ANCHOR_MIDDLE,
+                                      velocity, Vector(0, gravity),
+                                      LAYER_OBJECTS + 3, true);
+  }
+
+  m_timer.start(m_delay);
 }
 
 void
-Root::deactivate()
+Root::draw(DrawingContext &context)
 {
-  remove_me();
-  // No dead-script required for deactivation.
+  BadGuy::draw(context);
+
+  if (!m_base_surface) return;
+
+  Vector pos = m_start_position;
+  switch (m_dir)
+  {
+    case Direction::UP:
+      pos.x -= m_sprite->get_current_hitbox_x_offset();
+      pos.y -= get_bbox().get_height() + 17.5f;
+      break;
+
+    case Direction::DOWN:
+      pos.x -= m_sprite->get_current_hitbox_x_offset();
+      pos.y += get_bbox().get_height() - 5;
+      break;
+
+    case Direction::LEFT:
+      pos.x -= get_bbox().get_width() + 12.5f;
+      pos.y -= m_sprite->get_current_hitbox_y_offset();
+      break;
+
+    case Direction::RIGHT:
+      pos.x += get_bbox().get_width() - 5;
+      pos.y -= m_sprite->get_current_hitbox_y_offset();
+      break;
+
+    default: assert(false); break;
+  }
+  context.color().draw_surface(m_base_surface,
+                               pos,
+                               m_sprite->get_angle(),
+                               m_sprite->get_color(),
+                               m_sprite->get_blend(),
+                               m_layer+1);
 }
 
 void
 Root::active_update(float dt_sec)
 {
-  if (mystate == STATE_APPEARING) {
-    if (base_sprite->animation_done()) {
-      hatch_timer.start(HATCH_TIME);
-      mystate = STATE_HATCHING;
-    }
-  }
-  if (mystate == STATE_HATCHING) {
-    if (!hatch_timer.started()) mystate = STATE_GROWING;
-  }
-  else if (mystate == STATE_GROWING) {
-    offset_y -= dt_sec * SPEED_GROW;
-    if (offset_y < static_cast<float>(-m_sprite->get_height())) {
-      offset_y = static_cast<float>(-m_sprite->get_height());
-      mystate = STATE_SHRINKING;
+  BadGuy::active_update(dt_sec);
+
+  switch (m_state)
+  {
+    case STATE_HATCHING:
+      if (m_timer.check())
+      {
+        m_state = STATE_APPEARING;
+        m_timer.start(APPEAR_TIME);
+      }
+      break;
+
+    case STATE_APPEARING:
+    {
+      float size = 0;
+      switch (m_dir)
+      {
+        case Direction::LEFT:
+        case Direction::RIGHT:
+          size = get_bbox().get_width();
+          break;
+
+        case Direction::UP:
+        case Direction::DOWN:
+          size = get_bbox().get_height();
+          break;
+
+        default: assert(false); break;
+      }
```

Comment:
This `switch` can be replaced with ternary operator:
```cpp
const float size = (m_dir == Direction::LEFT || m_dir == Direction::RIGHT) ? get_bbox().get_width() : get_bbox().get_height();
```

Reply:
Won't that get quite long? I think this code is more readable.

Now what i could do is move it into a function

Reply:
The one-liner should be fine. You could also have it on two lines as well.